File: go/internal/navigator/controller.go
package navigator

import "strings"

// FilterController is the query/app-filter/bell-filter slice of the controller interface.
type FilterController interface {
	Query() string
	AppIcons() []WindowInfo
	AppFilterIndex() *int
	BellFilter() bool
	AppFilter() *string
	SetQuery(text string)
	SetDesktopNums(nums map[int]struct{})
	ToggleBellFilter()
	CycleAppFilter(direction int)
	ClearAppFilter()
}

// NavigationController is the selection/movement slice of the controller interface.
type NavigationController interface {
	SelectionIndex() int
	FlatList() []FlatItem
	FilteredWindows() []WindowInfo
	MoveUp()
	MoveDown()
	MovePageUp(pageSize int)
	MovePageDown(pageSize int)
	MoveToFirst()
	MoveToLast()
	SelectedItem() *FlatItem
	SelectedHWND() *uintptr
}

// TabController is the UIA tab discovery/expansion slice of the controller interface.
type TabController interface {
	SetTabs(hwnd uintptr, tabs []TabInfo)
	ToggleAllExpansions()
	TabCount(hwnd uintptr) int
	IsExpanded(hwnd uintptr) bool
}

// OverlayController is the full controller interface consumed by the overlay.
type OverlayController interface {
	FilterController
	NavigationController
	TabController
	Reset(windows []WindowInfo)
}

// FlatItem is a union of WindowInfo and TabInfo for the flat rendered list.
// Exactly one of Window/Tab is non-nil.
type FlatItem struct {
	Window *WindowInfo
	Tab    *TabInfo
}

// HWND returns the hwnd: the window's own hwnd, or the tab's parent hwnd.
func (f FlatItem) HWND() uintptr {
	if f.Window != nil {
		return f.Window.HWND
	}
	return f.Tab.HWND
}

// Controller implements OverlayController — pure Go, no UI dependency.
type Controller struct {
	AllWindows     []WindowInfo
	query          string
	desktopNums    map[int]struct{}
	appFilter      *string
	bellFilter     bool
	tabs           map[uintptr][]TabInfo
	expanded       map[uintptr]struct{}
	wantAllExpand  bool
	selectionIndex int
}

// NewController creates a fresh Controller for the given window list.
func NewController(windows []WindowInfo) *Controller {
	sel := -1
	if len(windows) > 0 {
		sel = 0
	}
	return &Controller{
		AllWindows:     append([]WindowInfo(nil), windows...),
		desktopNums:    make(map[int]struct{}),
		tabs:           make(map[uintptr][]TabInfo),
		expanded:       make(map[uintptr]struct{}),
		selectionIndex: sel,
	}
}

// ---------------------------------------------------------------------------
// FilterController
// ---------------------------------------------------------------------------

func (c *Controller) Query() string { return c.query }

func (c *Controller) BellFilter() bool { return c.bellFilter }

func (c *Controller) AppFilter() *string { return c.appFilter }

// AppIcons returns one representative WindowInfo per unique ProcessName in
// TextFilteredWindows, preserving recency (z-order) of first occurrence.
func (c *Controller) AppIcons() []WindowInfo {
	seen := make(map[string]struct{})
	var result []WindowInfo
	for _, w := range c.TextFilteredWindows() {
		if _, ok := seen[w.ProcessName]; !ok {
			seen[w.ProcessName] = struct{}{}
			result = append(result, w)
		}
	}
	return result
}

// AppFilterIndex returns a pointer to the index of the active app filter in AppIcons,
// or nil if no filter is active.
func (c *Controller) AppFilterIndex() *int {
	if c.appFilter == nil {
		return nil
	}
	for i, w := range c.AppIcons() {
		if w.ProcessName == *c.appFilter {
			return &i
		}
	}
	return nil
}

func (c *Controller) SetQuery(text string) {
	c.query = text
	if c.appFilter != nil {
		names := make(map[string]struct{})
		for _, w := range c.TextFilteredWindows() {
			names[w.ProcessName] = struct{}{}
		}
		hwndToWindow := make(map[uintptr]WindowInfo)
		for _, w := range c.AllWindows {
			hwndToWindow[w.HWND] = w
		}
		for h := range c.tabQueryMatches() {
			if w, ok := hwndToWindow[h]; ok {
				names[w.ProcessName] = struct{}{}
			}
		}
		if _, ok := names[*c.appFilter]; !ok {
			c.appFilter = nil
		}
	}
	flat := c.FlatList()
	if len(flat) > 0 {
		c.selectionIndex = 0
	} else {
		c.selectionIndex = -1
	}
}

func (c *Controller) SetDesktopNums(nums map[int]struct{}) {
	c.desktopNums = nums
	if c.appFilter != nil {
		names := make(map[string]struct{})
		for _, w := range c.TextFilteredWindows() {
			names[w.ProcessName] = struct{}{}
		}
		hwndToWindow := make(map[uintptr]WindowInfo)
		for _, w := range c.AllWindows {
			hwndToWindow[w.HWND] = w
		}
		for h := range c.tabQueryMatches() {
			if w, ok := hwndToWindow[h]; ok {
				names[w.ProcessName] = struct{}{}
			}
		}
		if _, ok := names[*c.appFilter]; !ok {
			c.appFilter = nil
		}
	}
	c.resetSelection()
}

func (c *Controller) ToggleBellFilter() {
	c.bellFilter = !c.bellFilter
	c.resetSelection()
}

func (c *Controller) CycleAppFilter(direction int) {
	icons := c.AppIcons()
	if len(icons) == 0 {
		return
	}
	idx := c.AppFilterIndex()
	var newIdx int
	if idx == nil {
		if direction > 0 {
			newIdx = 0
		} else {
			newIdx = len(icons) - 1
		}
	} else {
		newIdx = ((*idx) + direction + len(icons)) % len(icons)
	}
	name := icons[newIdx].ProcessName
	c.appFilter = &name
	c.resetSelection()
}

func (c *Controller) ClearAppFilter() {
	c.appFilter = nil
	c.resetSelection()
}

// ---------------------------------------------------------------------------
// Derived views
// ---------------------------------------------------------------------------

// TextFilteredWindows returns windows matching desktopNums + text query only (no app filter).
func (c *Controller) TextFilteredWindows() []WindowInfo {
	return FilterWindows(c.AllWindows, c.query, c.desktopNums)
}

// tabQueryMatches returns a map of hwnd → matching tabs for windows that are
// expanded but not matched by title/process name, when the query has text.
// Returns empty map when nothing is expanded.
func (c *Controller) tabQueryMatches() map[uintptr][]TabInfo {
	result := make(map[uintptr][]TabInfo)
	if len(c.expanded) == 0 {
		return result
	}
	if strings.TrimSpace(c.query) == "" {
		return result
	}
	tokens := strings.Fields(strings.ToLower(c.query))
	titleHWNDs := make(map[uintptr]struct{})
	for _, w := range c.TextFilteredWindows() {
		titleHWNDs[w.HWND] = struct{}{}
	}
	hwndToWindow := make(map[uintptr]WindowInfo)
	for _, w := range c.AllWindows {
		hwndToWindow[w.HWND] = w
	}
	for hwnd, tabs := range c.tabs {
		if _, ok := titleHWNDs[hwnd]; ok {
			continue
		}
		w, ok := hwndToWindow[hwnd]
		if !ok {
			continue
		}
		if len(c.desktopNums) > 0 {
			if _, ok := c.desktopNums[w.DesktopNumber]; !ok {
				continue
			}
		}
		var matching []TabInfo
		for _, t := range tabs {
			name := strings.ToLower(t.Name)
			allMatch := true
			for _, tok := range tokens {
				if !strings.Contains(name, tok) {
					allMatch = false
					break
				}
			}
			if allMatch {
				matching = append(matching, t)
			}
		}
		if len(matching) > 0 {
			result[hwnd] = matching
		}
	}
	return result
}

// FilteredWindows returns windows matching query AND all active filters.
func (c *Controller) FilteredWindows() []WindowInfo {
	titleHWNDs := make(map[uintptr]struct{})
	for _, w := range c.TextFilteredWindows() {
		titleHWNDs[w.HWND] = struct{}{}
	}
	tabMatches := c.tabQueryMatches()
	var result []WindowInfo
	for _, w := range c.AllWindows {
		_, inTitle := titleHWNDs[w.HWND]
		_, inTab := tabMatches[w.HWND]
		if !inTitle && !inTab {
			continue
		}
		if c.bellFilter && !w.HasNotification {
			continue
		}
		if c.appFilter != nil && w.ProcessName != *c.appFilter {
			continue
		}
		result = append(result, w)
	}
	return result
}

// FlatList returns filtered windows with expanded tab rows interleaved after their parent.
func (c *Controller) FlatList() []FlatItem {
	tabMatches := c.tabQueryMatches()
	var result []FlatItem
	for _, w := range c.FilteredWindows() {
		wCopy := w
		result = append(result, FlatItem{Window: &wCopy})
		if tabs, ok := tabMatches[w.HWND]; ok {
			for _, t := range tabs {
				tCopy := t
				result = append(result, FlatItem{Tab: &tCopy})
			}
		} else if _, expanded := c.expanded[w.HWND]; expanded {
			if tabs, ok := c.tabs[w.HWND]; ok {
				for _, t := range tabs {
					tCopy := t
					result = append(result, FlatItem{Tab: &tCopy})
				}
			}
		}
	}
	return result
}

// ---------------------------------------------------------------------------
// NavigationController
// ---------------------------------------------------------------------------

func (c *Controller) SelectionIndex() int { return c.selectionIndex }

func (c *Controller) MoveUp() {
	if c.selectionIndex > 0 {
		c.selectionIndex--
	}
}

func (c *Controller) MoveDown() {
	if c.selectionIndex < len(c.FlatList())-1 {
		c.selectionIndex++
	}
}

func (c *Controller) MovePageUp(pageSize int) {
	if c.selectionIndex > 0 {
		c.selectionIndex = max(0, c.selectionIndex-pageSize)
	}
}

func (c *Controller) MovePageDown(pageSize int) {
	last := len(c.FlatList()) - 1
	if c.selectionIndex < last {
		c.selectionIndex = min(last, c.selectionIndex+pageSize)
	}
}

func (c *Controller) MoveToFirst() {
	if len(c.FlatList()) > 0 {
		c.selectionIndex = 0
	}
}

func (c *Controller) MoveToLast() {
	last := len(c.FlatList()) - 1
	if last >= 0 {
		c.selectionIndex = last
	}
}

func (c *Controller) SelectedItem() *FlatItem {
	flat := c.FlatList()
	if c.selectionIndex < 0 || c.selectionIndex >= len(flat) {
		return nil
	}
	item := flat[c.selectionIndex]
	return &item
}

func (c *Controller) SelectedHWND() *uintptr {
	item := c.SelectedItem()
	if item == nil {
		return nil
	}
	h := item.HWND()
	return &h
}

// ---------------------------------------------------------------------------
// TabController
// ---------------------------------------------------------------------------

func (c *Controller) TabCount(hwnd uintptr) int {
	return len(c.tabs[hwnd])
}

func (c *Controller) IsExpanded(hwnd uintptr) bool {
	_, ok := c.expanded[hwnd]
	return ok
}

func (c *Controller) SetTabs(hwnd uintptr, tabs []TabInfo) {
	c.tabs[hwnd] = tabs
	if c.wantAllExpand && len(tabs) > 1 {
		c.expanded[hwnd] = struct{}{}
	}
}

// ToggleExpansion expands or collapses the tab list for hwnd.
// Only expands when the window has more than one tab.
// Always leaves selection on the parent window row.
func (c *Controller) ToggleExpansion(hwnd uintptr) {
	if _, ok := c.expanded[hwnd]; ok {
		delete(c.expanded, hwnd)
	} else if tabs, ok := c.tabs[hwnd]; ok && len(tabs) > 1 {
		c.expanded[hwnd] = struct{}{}
	}
	for i, item := range c.FlatList() {
		if item.Window != nil && item.Window.HWND == hwnd {
			c.selectionIndex = i
			break
		}
	}
}

func (c *Controller) ToggleAllExpansions() {
	hwndsWithTabs := make(map[uintptr]struct{})
	for _, w := range c.FilteredWindows() {
		if tabs, ok := c.tabs[w.HWND]; ok && len(tabs) > 1 {
			hwndsWithTabs[w.HWND] = struct{}{}
		}
	}
	if len(hwndsWithTabs) == 0 {
		c.wantAllExpand = !c.wantAllExpand
		return
	}
	// Check if any in hwndsWithTabs are currently expanded
	anyExpanded := false
	for hwnd := range hwndsWithTabs {
		if _, ok := c.expanded[hwnd]; ok {
			anyExpanded = true
			break
		}
	}
	if anyExpanded {
		for hwnd := range hwndsWithTabs {
			delete(c.expanded, hwnd)
		}
		c.wantAllExpand = false
	} else {
		for hwnd := range hwndsWithTabs {
			c.expanded[hwnd] = struct{}{}
		}
		c.wantAllExpand = true
	}
}

// ---------------------------------------------------------------------------
// Reset
// ---------------------------------------------------------------------------

func (c *Controller) Reset(windows []WindowInfo) {
	c.AllWindows = append([]WindowInfo(nil), windows...)
	c.query = ""
	c.desktopNums = make(map[int]struct{})
	c.appFilter = nil
	c.bellFilter = false
	c.tabs = make(map[uintptr][]TabInfo)
	c.expanded = make(map[uintptr]struct{})
	c.wantAllExpand = false
	if len(windows) > 0 {
		c.selectionIndex = 0
	} else {
		c.selectionIndex = -1
	}
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

func (c *Controller) resetSelection() {
	flat := c.FlatList()
	if len(flat) > 0 {
		c.selectionIndex = 0
	} else {
		c.selectionIndex = -1
	}
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: python/src/windows_navigator/controller.py
"""Pure-Python state machine for the overlay — no Tk dependency, fully testable."""

from __future__ import annotations

import functools
from typing import Protocol

from windows_navigator.filter import filter_windows
from windows_navigator.models import TabInfo, WindowInfo


class FilterControllerProtocol(Protocol):
    """Query text, desktop filter, app filter, and bell filter state."""

    query: str

    @property
    def desktop_nums(self) -> set[int]: ...
    @property
    def app_icons(self) -> list[WindowInfo]: ...
    @property
    def app_filter_index(self) -> int | None: ...
    @property
    def bell_filter(self) -> bool: ...
    @property
    def app_filter(self) -> str | None: ...

    def set_query(self, text: str) -> None: ...
    def set_desktop_nums(self, nums: set[int]) -> None: ...
    def toggle_bell_filter(self) -> None: ...
    def cycle_app_filter(self, direction: int) -> None: ...
    def clear_app_filter(self) -> None: ...


class NavigationControllerProtocol(Protocol):
    """Selection index and movement through the flat window/tab list."""

    selection_index: int

    @property
    def flat_list(self) -> list[WindowInfo | TabInfo]: ...
    @property
    def filtered_windows(self) -> list[WindowInfo]: ...

    def move_up(self) -> None: ...
    def move_down(self) -> None: ...
    def move_page_up(self, page_size: int) -> None: ...
    def move_page_down(self, page_size: int) -> None: ...
    def move_to_first(self) -> None: ...
    def move_to_last(self) -> None: ...
    def selected_item(self) -> WindowInfo | TabInfo | None: ...
    def selected_hwnd(self) -> int | None: ...


class TabControllerProtocol(Protocol):
    """UIA tab discovery and per-window expansion state."""

    def set_tabs(self, hwnd: int, tabs: list[TabInfo]) -> None: ...
    def toggle_all_expansions(self) -> None: ...
    def tab_count(self, hwnd: int) -> int: ...
    def is_expanded(self, hwnd: int) -> bool: ...


class OverlayControllerProtocol(
    FilterControllerProtocol,
    NavigationControllerProtocol,
    TabControllerProtocol,
    Protocol,
):
    """Full controller interface consumed by NavigatorOverlay.

    Composes the three focused sub-protocols.  Depend on a sub-protocol when
    only a subset of the interface is needed; depend on this when everything is.
    """

    def reset(self, windows: list[WindowInfo]) -> None: ...


class OverlayController:
    """Tracks query, filtered window list, tab expansions, and selection index."""

    def __init__(self, windows: list[WindowInfo]) -> None:
        self.all_windows = list(windows)
        self.query = ""
        self._desktop_nums: set[int] = set()
        self._app_filter: str | None = None
        self._bell_filter: bool = False
        self._tabs: dict[int, list[TabInfo]] = {}
        self._expanded: set[int] = set()
        self._want_all_expanded: bool = False
        self.selection_index = 0 if windows else -1

    # ------------------------------------------------------------------
    # Derived views
    # ------------------------------------------------------------------

    @property
    def desktop_nums(self) -> set[int]:
        return self._desktop_nums

    @functools.cached_property
    def text_filtered_windows(self) -> list[WindowInfo]:
        """Windows matching desktop badge filter + text query only (no app filter)."""
        return filter_windows(self.all_windows, self.query, self._desktop_nums or None)

    def _invalidate_text_filter_cache(self) -> None:
        self.__dict__.pop("text_filtered_windows", None)

    @property
    def _tab_query_matches(self) -> dict[int, list[TabInfo]]:
        """Expanded windows not matched by title/process but having tabs matching the text query.

        Returns empty dict when nothing is expanded — tabs are only searched in expanded state.
        """
        if not self._expanded:
            return {}
        if not self.query.strip():
            return {}
        tokens = self.query.casefold().split()
        title_hwnds = {w.hwnd for w in self.text_filtered_windows}
        hwnd_to_window = {w.hwnd: w for w in self.all_windows}
        result: dict[int, list[TabInfo]] = {}
        for hwnd, tabs in self._tabs.items():
            if hwnd in title_hwnds:
                continue
            w = hwnd_to_window.get(hwnd)
            if w is None:
                continue
            if self._desktop_nums and w.desktop_number not in self._desktop_nums:
                continue
            matching = [t for t in tabs if all(tok in t.name.casefold() for tok in tokens)]
            if matching:
                result[hwnd] = matching
        return result

    @property
    def filtered_windows(self) -> list[WindowInfo]:
        """Windows matching query AND all active filters (bell, app, and tab-query matches)."""
        title_hwnds = {w.hwnd for w in self.text_filtered_windows}
        tab_matches = self._tab_query_matches
        result: list[WindowInfo] = []
        for w in self.all_windows:
            if w.hwnd not in title_hwnds and w.hwnd not in tab_matches:
                continue
            if self._bell_filter and not w.has_notification:
                continue
            if self._app_filter is not None and w.process_name != self._app_filter:
                continue
            result.append(w)
        return result

    @property
    def flat_list(self) -> list[WindowInfo | TabInfo]:
        """Filtered windows with expanded tab rows interleaved immediately after their parent."""
        tab_matches = self._tab_query_matches
        result: list[WindowInfo | TabInfo] = []
        for w in self.filtered_windows:
            result.append(w)
            hwnd = w.hwnd
            if hwnd in tab_matches:
                result.extend(tab_matches[hwnd])
            elif hwnd in self._expanded and hwnd in self._tabs:
                result.extend(self._tabs[hwnd])
        return result

    @property
    def app_icons(self) -> list[WindowInfo]:
        """One representative WindowInfo per unique process_name in text_filtered_windows.

        Preserves the recency (z-order) of the first occurrence of each process.
        """
        seen: set[str] = set()
        result: list[WindowInfo] = []
        for w in self.text_filtered_windows:
            if w.process_name not in seen:
                seen.add(w.process_name)
                result.append(w)
        return result

    @property
    def app_filter_index(self) -> int | None:
        """Index of the selected process in app_icons, or None if no filter active."""
        if self._app_filter is None:
            return None
        for i, w in enumerate(self.app_icons):
            if w.process_name == self._app_filter:
                return i
        return None

    @property
    def bell_filter(self) -> bool:
        return self._bell_filter

    @property
    def app_filter(self) -> str | None:
        return self._app_filter

    # ------------------------------------------------------------------
    # Tab management
    # ------------------------------------------------------------------

    def tab_count(self, hwnd: int) -> int:
        """Number of UIA tabs fetched for *hwnd* (0 if none fetched yet)."""
        return len(self._tabs.get(hwnd, []))

    def is_expanded(self, hwnd: int) -> bool:
        """True if *hwnd*'s tab list is currently expanded in the overlay."""
        return hwnd in self._expanded

    def set_tabs(self, hwnd: int, tabs: list[TabInfo]) -> None:
        """Store fetched tab data for *hwnd*. Called from the main thread after background fetch."""
        self._tabs[hwnd] = tabs
        if self._want_all_expanded and len(tabs) > 1:
            self._expanded.add(hwnd)

    def toggle_expansion(self, hwnd: int) -> None:
        """Expand or collapse the tab list for *hwnd*.

        Only expands when the window has more than one tab. Always leaves
        selection on the parent window row.
        """
        if hwnd in self._expanded:
            self._expanded.discard(hwnd)
        elif hwnd in self._tabs and len(self._tabs[hwnd]) > 1:
            self._expanded.add(hwnd)
        flat = self.flat_list
        for i, item in enumerate(flat):
            if isinstance(item, WindowInfo) and item.hwnd == hwnd:
                self.selection_index = i
                break

    def toggle_all_expansions(self) -> None:
        """Expand all visible windows that have tabs, or collapse all if any are expanded.

        If tabs have not arrived yet, sets a flag so each window auto-expands when
        its tabs are fetched.  Pressing again while the flag is set cancels the intent.
        """
        hwnds_with_tabs = {
            w.hwnd for w in self.filtered_windows
            if w.hwnd in self._tabs and len(self._tabs[w.hwnd]) > 1
        }
        if not hwnds_with_tabs:
            # Only flip the deferred flag when some filtered windows still have tabs loading.
            # If all filtered windows already have tab data (just single-tab), do nothing —
            # flipping would corrupt intent state without any visible effect.
            if any(w.hwnd not in self._tabs for w in self.filtered_windows):
                self._want_all_expanded = not self._want_all_expanded
            return
        if self._expanded & hwnds_with_tabs:
            self._expanded.clear()
            self._want_all_expanded = False
        else:
            self._expanded |= hwnds_with_tabs
            self._want_all_expanded = True
        n = len(self.flat_list)
        if n == 0:
            self.selection_index = -1
        elif self.selection_index >= n:
            self.selection_index = n - 1

    def selected_item(self) -> WindowInfo | TabInfo | None:
        """Return the currently selected item (window row or tab row), or None."""
        flat = self.flat_list
        if self.selection_index < 0 or self.selection_index >= len(flat):
            return None
        return flat[self.selection_index]

    # ------------------------------------------------------------------
    # Mutation helpers
    # ------------------------------------------------------------------

    def set_desktop_nums(self, nums: set[int]) -> None:
        """Update desktop badge filter and reset selection to the first visible row.

        If the active app filter no longer appears in the new filtered list,
        it is automatically cleared.
        """
        self._desktop_nums = nums
        self._invalidate_text_filter_cache()
        if self._app_filter is not None:
            names = {w.process_name for w in self.text_filtered_windows}
            hwnd_to_window = {w.hwnd: w for w in self.all_windows}
            names |= {
                hwnd_to_window[h].process_name
                for h in self._tab_query_matches
                if h in hwnd_to_window
            }
            if self._app_filter not in names:
                self._app_filter = None
        self.selection_index = 0 if self.flat_list else -1

    def set_query(self, text: str) -> None:
        """Update text filter query and reset selection to the first visible row.

        If the active app filter no longer appears in the new text-filtered list,
        it is automatically cleared.
        """
        self.query = text
        self._invalidate_text_filter_cache()
        if self._app_filter is not None:
            names = {w.process_name for w in self.text_filtered_windows}
            hwnd_to_window = {w.hwnd: w for w in self.all_windows}
            names |= {
                hwnd_to_window[h].process_name
                for h in self._tab_query_matches
                if h in hwnd_to_window
            }
            if self._app_filter not in names:
                self._app_filter = None
        self.selection_index = 0 if self.flat_list else -1

    def cycle_app_filter(self, direction: int) -> None:
        """Advance the app filter selection by *direction* (+1 or -1) with wrap."""
        icons = self.app_icons
        if not icons:
            return
        current = self.app_filter_index
        if current is None:
            new_index = 0 if direction > 0 else len(icons) - 1
        else:
            new_index = (current + direction) % len(icons)
        self._app_filter = icons[new_index].process_name
        self.selection_index = 0 if self.flat_list else -1

    def clear_app_filter(self) -> None:
        """Remove the active app filter and reset selection."""
        self._app_filter = None
        self.selection_index = 0 if self.flat_list else -1

    def toggle_bell_filter(self) -> None:
        """Toggle the notification-only filter and reset selection."""
        self._bell_filter = not self._bell_filter
        self.selection_index = 0 if self.flat_list else -1

    def move_up(self) -> None:
        n = len(self.flat_list)
        if n == 0:
            return
        if self.selection_index <= 0:
            self.selection_index = n - 1
        else:
            self.selection_index -= 1

    def move_down(self) -> None:
        n = len(self.flat_list)
        if n == 0:
            return
        if self.selection_index >= n - 1:
            self.selection_index = 0
        else:
            self.selection_index += 1

    def move_page_up(self, page_size: int) -> None:
        if self.selection_index > 0:
            self.selection_index = max(0, self.selection_index - page_size)

    def move_page_down(self, page_size: int) -> None:
        last = len(self.flat_list) - 1
        if self.selection_index < last:
            self.selection_index = min(last, self.selection_index + page_size)

    def move_to_first(self) -> None:
        if self.flat_list:
            self.selection_index = 0

    def move_to_last(self) -> None:
        last = len(self.flat_list) - 1
        if last >= 0:
            self.selection_index = last

    def reset(self, windows: list[WindowInfo]) -> None:
        """Reset all state for a fresh show() call."""
        self.all_windows = list(windows)
        self.query = ""
        self._desktop_nums = set()
        self._app_filter = None
        self._bell_filter = False
        self._tabs = {}
        self._expanded = set()
        self._want_all_expanded = False
        self.selection_index = 0 if windows else -1
        self._invalidate_text_filter_cache()

    # ------------------------------------------------------------------
    # Read helpers
    # ------------------------------------------------------------------

    def selected_hwnd(self) -> int | None:
        """Return the hwnd of the currently selected item (window or tab's parent), or None."""
        item = self.selected_item()
        return item.hwnd if item is not None else None
